llvm/lib/Transforms/Utils/CallPromotionUtils.cpp

```diff
@@ -712,7 +712,12 @@ bool llvm::tryPromoteCall(CallBase &CB) {
       VTablePtrLoad, VTablePtrLoad->getParent(), BBI, 0, nullptr, nullptr);
   if (!VTablePtr)
     return false; // No vtable found.
-  APInt VTableOffsetGVBase(DL.getTypeSizeInBits(VTablePtr->getType()), 0);
+
+  // VTablePtr may not actually have a pointer type, so we can't find the
+  // address space through it. However, we know that the address space must be
+  // the same one used to load from the vtable in VTablePtrLoad.
```

Comment:
I wouldn't rely on that, e.g. https://reviews.llvm.org/D114533 wanted to relax that restriction in some cases. Also, VTablePtrLoad->getPointerAddressSpace() is the address space of the pointer to the vtable pointer, not the vtable pointer itself (which would be VTablePtrLoad->getType()->getPointerAddressSpace()). At least I think what I said is correct, I confused myself several times writing this.

That is, we're dealing with:

```
  %object_base = alloca ptr
  %object = %object_base

  %vtable_ptr = getelementptr i8, ptr @gv, iN %vtable_gv_offset
  store ptr %vtable_ptr, ptr %object

  %vtable_ptr_load = load ptr, ptr %object
  %vtable_base_ptr = %vtable_ptr_load
  %vtable_entry_ptr = getelementptr i8, ptr %vtable_base_ptr, iN %vtable_offset
  %vtable_entry_load = load ptr, ptr %vtable_entry_ptr
  %callee = %vtable_entry_load
  call ptr %callee
```

Where we want vtable_ptr's address space, not object's.

Reply:
I'd just add || !VTablePtr->isPointerType() to line 713

Reply:
Revised to implement this fix instead.